Make `jl_timer_block_t` allocation separate from timer-start (JL_TIMING)

topolarity opened the pull request:
This includes several changes to the TIMING API:
  - Adds `JL_TIMING_CREATE_BLOCK(block, subsystem, event)` to create a timing block _without_ starting it
  - Adds `jl_timing_block_start` to start a timing block which was created with JL_TIMING_CREATE_BLOCK
  - Removes the C++-specific RAII implementation for JL_TIMING. Although it'd be nice to support JL_TIMING without GCC/Clang, the reality is that the C API prevents that from being achievable.
  - Renames JL_TIMING_CURRENT_BLOCK to JL_TIMING_DEFAULT_BLOCK

In effect JL_TIMING(subsystem, event) is now equivalent to:

```
JL_TIMING_CREATE(__timing_block, subsystem, event);
jl_timing_block_start(&__timing_block);
```

which also means that conditional events can be supported with:

```
JL_TIMING_CREATE(__timing_block, subsystem, event);
if (condition)
    jl_timing_block_start(&__timing_block);
```

topolarity committed:
timing: Introduce `JL_TIMING_CREATE_BLOCK` to separate alloc/init

This includes several changes to the TIMING API:
  - Adds `JL_TIMING_CREATE_BLOCK(block, subsystem, event)` to create
    a timing block _without_ starting it
  - Adds `jl_timing_block_start` to start a timing block which was
    created with JL_TIMING_CREATE_BLOCK
  - Removes the C++-specific RAII implementation for JL_TIMING. Although
    it'd be nice to support JL_TIMING without GCC/Clang, the reality is
    that the C API prevents that from being achievable.
  - Renames `JL_TIMING_CURRENT_BLOCK` to `JL_TIMING_DEFAULT_BLOCK`

To summarize, `JL_TIMING(subsystem, event)` is now equivalent to:

```
JL_TIMING_CREATE(__timing_block, subsystem, event);
jl_timing_block_start(&__timing_block);
```

which also means that conditional events can be supported with:

```
JL_TIMING_CREATE(__timing_block, subsystem, event);
if (condition)
    jl_timing_block_start(&__timing_block);
```

topolarity commented:
@pchintalapudi  This is an alternative to the approach taken in https://github.com/JuliaLang/julia/pull/49532 I wanted to take a stab in this direction and see if it felt a bit less rigid.

Please let me know if you are a fan of the new approach, or if you find the new macros difficult to reason about for any reason

topolarity commented:
Also I haven't had a chance to test this with VTune yet, but that's definitely a requirement before this is ready to merge

pchintalapudi commented:
I'd be interested to see usage of the api, and also how it handles conditional SWEEP_FULL and SWEEP_INCREMENTAL events. Imo we should also try to remove all the ifdef USE_TRACY blocks in favor of handling coloring directly in the timing code (maybe a Tracy colors array?) since we won't need to do conditional coloring anymore.

topolarity committed:
Split GC_Sweep JL_TIMING event into incremental/full versions

topolarity commented:
> I am confused. It seems like you have deleted the standard C++-specific RAII, in favor of only having the special support for C exposed by GCC and Clang?

Yes, that's right

topolarity commented:
> I'd be interested to see usage of the api, and also how it handles conditional SWEEP_FULL and SWEEP_INCREMENTAL events.

Sure thing - I added https://github.com/JuliaLang/julia/pull/49647/commits/9b5ca5158436df6f5fb06f3adbe3087af018eaa8 as an example

Another side effect of this PR btw is that the ITTAPI and TRACY internal macros are a bit more similar now. Since `TracyCZoneN` has been manually unfolded into timing.h, `_TRACY_START` and `_TRACY_STOP` now exist and work like their ITTAPI counterparts

> Imo we should also try to remove all the ifdef USE_TRACY blocks in favor of handling coloring directly in the timing code (maybe a Tracy colors array?) since we won't need to do conditional coloring anymore.

I don't think I'm going to tackle that here in this PR, but I'm in agreement with the goal :+1:

topolarity committed:
timing: Create ITTAPI events on the fly

Instead of initializing all ITTAPI events during init, this change makes
ITTAPI events use a statically-allocated object to track whether the
event has been created.

This makes our generation of events more similar to the Tracy API, where
source locations are generated statically, in-line at each macro call-site
instead of constructing them all up front.

pchintalapudi commented:
In #49679 I make WITH_TIMING_COUNTS report both self and total time, so maybe we do want to keep the longer list of events for that?

topolarity committed:
Remove pointer indirection in `_TRACY_STOP`

topolarity committed:
Reset `active` status for timing zone upon block entry

pchintalapudi commented on src/gc.c:
```diff
@@ -3301,11 +3301,14 @@ static int _jl_gc_collect(jl_ptls_t ptls, jl_gc_collection_t collection)
+                               GC, GC_FullSweep);
+        jl_timing_block_start(sweep_full ? &full_timing_block : &incremental_timing_block);
 #ifdef USE_TRACY
         if (sweep_full) {
```
Can we remove this if?

topolarity replied:
Nice catch. We can - TracyCZoneColor has a guard: `if( !ctx.active ) return;`

pchintalapudi commented on src/timing.h:
```diff
@@ -4,6 +4,7 @@
 #define JL_TIMING_H
 
 #include "julia.h"
+#include <stdbool.h>
```
No need for stdbool, just use int (I don't believe we use stdbool anywhere else anyways).

topolarity replied:
I find `bool` a lot clearer for expressing intent. Plus, it behaves as you'd expect in comparisons

However you're right that this isn't used anywhere else so I'll bow to tradition here.

pchintalapudi reviewed (approved):
Looks good to me, modulo some nits.

KristofferC commented on src/timing.h:
```diff
@@ -319,52 +323,65 @@ STATIC_INLINE void _jl_timing_counts_destroy(jl_timing_counts_t *block) JL_NOTSA
+        uint64_t t = cycleclock(); (void)t;
+        _ITTAPI_STOP(block);
+        _COUNTS_STOP(&block->counts_ctx, t);
+        _TRACY_STOP(&block->tracy_ctx);
```
```suggestion
        _TRACY_STOP(block->tracy_ctx);
```
